### PR TITLE
[hail] clean up BlockMatrixMap IRs, delegate optimization of broadcasting for scalars to Simplify pass

### PR DESCRIPTION
**Pull request by catoverdrive**
The main goal of this PR was to remove some of the vector/scalar logic from the BlockMatrixMap2 node, and to support the scalar operations on BlockMatrixMap. I basically accomplished this by taking the cases that are matched on in BlockMatrixMap2 and lifting them into the Simplify rules.

The only endpoint that I believe I needed to cover was the BlockMatrix.pyExecute() one; all the others will go through the usual CompileAndEvaluate.

There's another part of the PR that fixes the variable bindings, which are currently hard-coded and unchecked. I needed this to construct the right expressions for the IR expressions, so I changed it to handle variable bindings with the rest of our IR.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala [hunk 1]
```diff
@@ -58,7 +60,12 @@ object BlockMatrixIR {
 abstract sealed class BlockMatrixIR extends BaseIR {
   def typ: BlockMatrixType
 
-  def pyExecute(): BlockMatrix = ExecuteContext.scoped(execute)
+  def pyExecute(): BlockMatrix = {
+    ExecuteContext.scoped { ctx =>
+      val opt = Optimize.optimize(this, noisy = true, context = "BlockMatrixIR", ctx = Some(ctx))
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala]: this should really go through a lowering pass -- there could totally be something like a TableAggregate in a value IR somewhere

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala]: Ok, sounds good! I'll just replace this a call to LoweringPipeline.relationalLowerer?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala]: Yeah. And maybe call the execute through Interpret the way we do for MatrixIR/TableIR that still go through this legacy path?

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala]: oh, good call. I looked for the analogous path but I couldn't find where it was called in python

**Pull request merged**